cmd/umoci.go

```diff
@@ -188,12 +189,79 @@ func tarUnpack(oci casext.Engine, tag string, bundlePath string, callback layer.
 	return umoci.Unpack(oci, tag, bundlePath, opts)
 }
 
+func which(name string) string {
+	return whichSearch(name, strings.Split(os.Getenv("PATH"), ":"))
+}
+
+func whichSearch(name string, paths []string) string {
+	var search []string
+
+	if strings.ContainsRune(name, os.PathSeparator) {
+		if path.IsAbs(name) {
+			search = []string{name}
+		} else {
+			search = []string{"./" + name}
+		}
+	} else {
+		search = []string{}
+		for _, p := range paths {
+			search = append(search, path.Join(p, name))
+		}
+	}
+
+	for _, fPath := range search {
+		if err := unix.Access(fPath, unix.X_OK); err == nil {
+			return fPath
+		}
+	}
+
+	return ""
+}
+
+func extractSingleSquash(squashFile string, extractDir string, rootless bool) error {
+	err := os.MkdirAll(extractDir, 0755)
+	if err != nil {
+		return err
+	}
+
+	var uCmd []string
+	if which("squashtool") != "" {
+		uCmd = []string{"squashtool", "extract", "--whiteouts", "--perms"}
+		if !rootless {
+			uCmd = append(uCmd, "--devs", "--sockets", "--owners")
```

Comment:
Doesn't this introduce correctness issues around permissions and such?

Reply:
It seems like we should probably run unsquashfs in a user+mount+net ns instead like we do with umoci, so it can write all of these kinds of things.

Reply:
It will not retain owners correctly. but i'm not sure how you could hope to do that as non-root.  perhaps if the process was executed in the user-namespace, and it used `--whiteouts --perms --owners` then it could do the right thing.

Reply:
I accept that this isn't perfect, but I'd suggest that if you want to ensure correctness, then the right thing to do is just fail, as unsquashfs can't offer correctness either.

Reply:
> It seems like we should probably run unsquashfs in a user+mount+net ns instead like we do with umoci, so it can write all of these kinds of things.

the issue with unsquashfs is that it does not understand whiteouts. Without explicit knowledge of them, you're destined to fail.  As you'll end up doing:

   * extract a layer with /some/dir
   * extract a layer with /some/dir/.whiteout

and unsquashfs will try to create the .whiteout file, which is a char device over the top of the existing 'dir' and will fail with "is a directory".

Reply:
so i guess there are 2 use cases, and I had only considered one.

 1. extract layer1, layer2, layer3 to extractDir/

    here the goal is to have extractDir/ have the same contents that overlay
    mounts would give you of layer3 over layer2 over layer1.


 2. extract layer1 to extractDir
    here the goal is to have exactly the contents of layer1


The difference really is that you want '--whiteouts' on 1 but not on '2'.  You have to know which one you're doing.

I had only considered '1'.